main.go

```diff
@@ -1,121 +1,68 @@
+		ret, err := rq.useCase.Execute()
 		if err != nil {
-			log.Println("get comments failed with error: ", err)
+			log.Println("execute usecase failed with error: ", err)
 			writer.WriteHeader(500)
 			return
 		}
 
-		// Combine and return response
-		postWithComments := combinePostWithComments(posts, comments)
-		resp := PostWithCommentsResponse{Posts: postWithComments}
-		buf, err := json.Marshal(resp)
+		// encoding data
+		buf, err := rq.formatter.Encode(ret)
 		if err != nil {
 			log.Println("unable to parse response: ", err)
 			writer.WriteHeader(500)
 		}
 
-		writer.Header().Set("Content-Type", "application/json")
+		// response to client
+		writer.Header().Set("Content-Type", rq.formatter.GetContentType())
 		_, err = writer.Write(buf)
 	})
-
-	log.Println("httpServer starts ListenAndServe at 8080")
-	log.Fatal(http.ListenAndServe(":8080", nil))
-}
-
-func getPosts() ([]Post, error) {
-	resp, err := http.Get(getPostsEndpoint)
-	if err != nil {
-		return nil, err
-	}
-	body, err := ioutil.ReadAll(resp.Body)
-	defer func() {
-		_ = resp.Body.Close()
-	}()
-
-	var posts []Post
-	if err = json.Unmarshal(body, &posts); err != nil {
-		return nil, err
-	}
-
-	return posts, nil
 }
 
-func getComments() ([]Comment, error) {
-	resp, err := http.Get(getCommentsEndpoint)
-	if err != nil {
-		return nil, err
-	}
-	body, err := ioutil.ReadAll(resp.Body)
-	defer func() {
-		_ = resp.Body.Close()
-	}()
-
-	var comments []Comment
-	if err = json.Unmarshal(body, &comments); err != nil {
-		return nil, err
-	}
+func newGetPostsWithCommentsUsecase() *usecase.GetPostsWithCommentsUsecase {
+	dataFormatter := &dataformatter.JSONDataFormatter{}
+	postData := data.NewPostData(dataFormatter)
+	getPostsUsecase := usecase.NewGetPostsUsecase(postData)
+	commentData := data.NewCommentData(dataFormatter)
+	getCommentsUsecase := usecase.NewGetCommentsUsecase(commentData)
 
-	return comments, nil
+	return usecase.NewGetPostsWithCommentsUsecase(getPostsUsecase, getCommentsUsecase)
 }
 
-func combinePostWithComments(posts []Post, comments []Comment) []PostWithComments {
-	commentsByPostID := map[int64][]Comment{}
-	for _, comment := range comments {
-		commentsByPostID[comment.PostID] = append(commentsByPostID[comment.PostID], comment)
-	}
-
-	result := make([]PostWithComments, 0, len(posts))
-	for _, post := range posts {
-		result = append(result, PostWithComments{
-			ID:       post.ID,
-			Title:    post.Title,
-			Comments: commentsByPostID[post.ID],
-		})
+func main() {
+	// define my mapper
+	myFormatter := &formatter.JSONFormatter{
+		ContentType: "application/json",
 	}
+	// define usecase
+	getPostsWithCommentsUsecase := newGetPostsWithCommentsUsecase()
+	// define handler
+	handleRequest(&requestHandler{
+		endPoint:  "/posts-with-comments",
+		useCase:   getPostsWithCommentsUsecase,
+		formatter: myFormatter,
+	})
 
-	return result
+	log.Println("httpServer starts ListenAndServe at 8080")
+	log.Fatal(http.ListenAndServe(":8080", nil))
 }
```

Comment:
Anyway, thanks for the submission, and sorry for the delay in code-review.
Happy coding!

Reply:
thanks for your reviews :D